gherkin/java/src/main/java/io/cucumber/gherkin/PickleCompiler.java

```diff
@@ -97,10 +101,19 @@ private void compileRule(List<Pickle> pickles, Rule rule, List<Tag> parentTags,
 
     private void compileScenario(List<Pickle> pickles, Scenario scenario, List<Tag> parentTags, List<Step> backgroundSteps, String language, String uri) {
         List<PickleStep> steps = new ArrayList<>();
-        if (!scenario.getSteps().isEmpty())
-            steps.addAll(pickleSteps(backgroundSteps));
+        if (!scenario.getSteps().isEmpty()) {
+            List<Step> allSteps = new ArrayList<>();
+            allSteps.addAll(backgroundSteps);
+            allSteps.addAll(scenario.getSteps());
 
-        steps.addAll(pickleSteps(scenario.getSteps()));
+            StepKeywordType lastKeywordType = StepKeywordType.UNKNOWN;
+            for (Step step : allSteps) {
+                if (step.getKeywordType() != StepKeywordType.CONJUNCTION)
```

Comment:
This logic is repeated 3 times in the class, which makes me think it may have a more natural home. Is the pickle the right place to resolve this, or is it the job of the consuming cucumber (which would do this in one place)?

Reply:
That's correct. I've thought about abstracting this out into a single place, but that's not so easy because of the fact that the "lastKeywordType" state needs to be maintained during the `compileScenario` and `compileScenarioOutline` functions.
I don't think this belongs in the consumers - even though they would indeed do this in one place - because the pickles are supposed to be directly executable, that is, it has all the necessary substitutions applied. While writing this code, I imagined this to be one of the substitutions the consumer needs to be able to rely on to have taken place.

Reply:
Maybe this should be something that's done in `compile` after `compileFeature` but before the pickles are returned?

Reply:
I looked at the options to do that. Although the pattern does repeat itself three times at an abstract level; on an implementation level, it's really two of one type and one of another; for now, I'll leave it as it is, because the deduplication is rather minimal and a repeated pattern will remain (2&1->1&1).